### PR TITLE
refactor: improve trie prefetching

### PR DESCRIPTION
**Pull request by pugachAG**
* Fix instances of `clippy::result_unit_err` warning
* Add explicit assert in tests on timeout while waiting for queue to become empty
* [Avoid single-letter variable names](https://near.github.io/nearcore/practices/style.html#standard-naming)
* Avoid code duplication in `prefetching.0.lock().expect(POISONED_LOCK_ERR)` by introducing helper `lock` function

### CODE REVIEW

File: core/store/src/trie/prefetching_trie_storage.rs [hunk 1]
```diff
@@ -39,6 +39,10 @@ const NUM_IO_THREADS: usize = 8;
 ///
 /// Each I/O threads will have its own copy of `TriePrefetchingStorage`, so
 /// this should remain a cheap object.
+///
+/// Please note that this should only be used from the background threads
+/// performing prefetching. The main thread uses `PrefetchStagingArea`
+/// directly to look up for prefetched values before going to the database.
```
Comment (akhi3030) [inline core/store/src/trie/prefetching_trie_storage.rs:45]:
```suggestion
/// directly to look up prefetched values before going to the database.
```

File: core/store/src/trie/prefetching_trie_storage.rs [hunk 2]
```diff
@@ -133,6 +137,18 @@ enum PrefetchSlot {
     Done(Arc<[u8]>),
 }
 
+impl PrefetchSlot {
+    /// Reserved memory capacity for a value from the prefetching area.
```
Comment (akhi3030) [inline core/store/src/trie/prefetching_trie_storage.rs:141]:
```suggestion
    /// Returns amount of memory reserved for a value in the prefetching area.
```

File: runtime/runtime/src/prefetch.rs [hunk 3]
```diff
@@ -85,7 +89,10 @@ impl TriePrefetcher {
     /// Start prefetching data for processing the receipts.
```
Comment (akhi3030) [inline runtime/runtime/src/prefetch.rs:89]: Not part of the PR but this might be worth some drive-by cleanup.

```suggestion
    /// Starts prefetching data for processing the receipts.
```

File: runtime/runtime/src/prefetch.rs [hunk 4]
```diff
@@ -85,7 +89,10 @@ impl TriePrefetcher {
...
     ///
     /// Returns an error if the prefetching queue is full.
```
Comment (akhi3030) [inline runtime/runtime/src/prefetch.rs:91]: This comment has the danger of becoming obsolete and is also not complete.  I propose something like below .

`Returns an error if prefetching for any receipt fails.  The function is not idempotent; in case of failure, prefetching for some receipts may have been initiated.`

File: runtime/runtime/src/prefetch.rs [hunk 5]
```diff
@@ -85,7 +89,10 @@ impl TriePrefetcher {
...
-    pub(crate) fn input_receipts(&mut self, receipts: &[Receipt]) -> Result<(), ()> {
+    pub(crate) fn input_receipts(
```
Comment (akhi3030) [inline runtime/runtime/src/prefetch.rs:92]: I think we should change the name of this function to something like `prefetch_data_for_receipts`.  But maybe not as part of this PR.  Up to you.

File: runtime/runtime/src/prefetch.rs [hunk 6]
```diff
@@ -122,7 +129,7 @@ impl TriePrefetcher {
     pub(crate) fn input_transactions(
```
Comment (akhi3030) [inline runtime/runtime/src/prefetch.rs:129]: I propose similar changes here as the above function.

File: runtime/runtime/src/prefetch.rs [hunk 7]
```diff
@@ -157,12 +164,12 @@ impl TriePrefetcher {
         self.prefetch_api.clear_data();
     }
 
-    fn prefetch_trie_key(&self, trie_key: TrieKey) -> Result<(), ()> {
+    fn prefetch_trie_key(&self, trie_key: TrieKey) -> Result<(), PrefetchQueueFullError> {
```
Comment (akhi3030) [inline runtime/runtime/src/prefetch.rs:167]: I propose that the error type returned here be something like `enum PrefetchTrieKeyError` and `QueueFull` be a variant of it.  That makes it easier to add additional error types in future with minimal changes.

File: runtime/runtime/src/prefetch.rs [hunk 8]
```diff
@@ -157,12 +164,12 @@ impl TriePrefetcher {
...
         let queue_full = self.prefetch_api.prefetch_trie_key(self.trie_root, trie_key).is_err();
```
Comment (akhi3030) [inline runtime/runtime/src/prefetch.rs:168]: If we make the change to error type above, then `self.prefetch_api.prefetch_trie_key()` should be returning the type.

**Review by akhi3030 — COMMENTED** [review]
I am not super familiar with the code but trying to get more into it.  So some drive-by comments.  Feel free to ignore if they do not make sense.

File: runtime/runtime/src/prefetch.rs (hunk 4 above)
Comment (pugachAG) [reply runtime/runtime/src/prefetch.rs:91]: lgtm, included

File: runtime/runtime/src/prefetch.rs (hunk 8 above)
Comment (pugachAG) [reply runtime/runtime/src/prefetch.rs]: good point, done!

File: runtime/runtime/src/lib.rs [hunk 9]
```diff
@@ -1169,7 +1169,7 @@ impl Runtime {
         let mut prefetcher = TriePrefetcher::new_if_enabled(trie.clone());
 
         if let Some(prefetcher) = &mut prefetcher {
-            let _queue_full = prefetcher.input_transactions(transactions);
+            _ = prefetcher.prefetch_transactions_data(transactions);
```
Comment (jakmeier) [inline runtime/runtime/src/lib.rs:1172]: With this change, it's slightly harder to understand why we ignore the result.

Optional: add a comment

File: runtime/runtime/src/lib.rs [hunk 10]
```diff
@@ -1286,7 +1286,7 @@ impl Runtime {
         // We first process local receipts. They contain staking, local contract calls, etc.
         if let Some(prefetcher) = &mut prefetcher {
             prefetcher.clear();
-            let _queue_full = prefetcher.input_receipts(&local_receipts);
+            _ = prefetcher.prefetch_receipts_data(&local_receipts);
```
Comment (jakmeier) [inline runtime/runtime/src/lib.rs:1289]: (dito) Optional: add a comment to explain why error is ignored

File: runtime/runtime/src/lib.rs [hunk 11]
```diff
@@ -1313,7 +1313,7 @@ impl Runtime {
 
             if let Some(prefetcher) = &mut prefetcher {
                 prefetcher.clear();
-                let _queue_full = prefetcher.input_receipts(std::slice::from_ref(&receipt));
+                _ = prefetcher.prefetch_receipts_data(std::slice::from_ref(&receipt));
```
Comment (jakmeier) [inline runtime/runtime/src/lib.rs:1316]: (dito) Optional: add a comment

File: runtime/runtime/src/lib.rs [hunk 12]
```diff
@@ -1336,7 +1336,7 @@ impl Runtime {
         // And then we process the new incoming receipts. These are receipts from other shards.
         if let Some(prefetcher) = &mut prefetcher {
             prefetcher.clear();
-            let _queue_full = prefetcher.input_receipts(&incoming_receipts);
+            _ = prefetcher.prefetch_receipts_data(&incoming_receipts);
```
Comment (jakmeier) [inline runtime/runtime/src/lib.rs:1339]: (dito) Optional: add a comment

File: runtime/runtime/src/prefetch.rs [hunk 13]
```diff
@@ -210,7 +221,11 @@ mod tests {
         test_utils::{create_test_store, test_populate_trie},
         ShardTries, ShardUId, Trie, TrieConfig,
     };
-    use std::{rc::Rc, str::FromStr, time::Duration};
+    use std::{
+        rc::Rc,
+        str::FromStr,
+        time::{Duration, Instant},
+    };
```
Comment (jakmeier) [inline runtime/runtime/src/prefetch.rs:228]: We don't group imports like this in nearcore: https://near.github.io/nearcore/practices/style.html#import-granularity

Best to update your local rust analyzer setting, then it will be done automatically.
We are not very strict about this though, you will find lots of (mostly old) code that groups it different than the style guide suggests.

File: runtime/runtime/src/prefetch.rs [hunk 14]
```diff
@@ -293,8 +308,7 @@ mod tests {
         let prefetch_keys = accounts_to_trie_keys(prefetch);
 
         let shard_uids = vec![ShardUId::single_shard()];
-        let mut trie_config = TrieConfig::default();
-        trie_config.enable_receipt_prefetching = true;
+        let trie_config = TrieConfig { enable_receipt_prefetching: true, ..TrieConfig::default() };
```
Comment (jakmeier) [inline runtime/runtime/src/prefetch.rs:311]: :heart:

File: runtime/runtime/src/prefetch.rs (hunk 7 above)
Comment (akhi3030) [reply runtime/runtime/src/prefetch.rs]: LGTM!

**Review by jakmeier — APPROVED** [review]
Looks great, happy to see the code improving! :smile: 

I left very few comments, mostly optional ones. Feel free to merge anytime you think it's ready.

___

_No action required but some thoughts:_

On a more code-philosophical note, I think `PrefetchError` is a positive change here but there are two reasons why it was like this.
- It is a common pattern in Rust code that a method that are designed to fail often under normal circumstances should return ownership of all arguments that it owns. This would allow to retry, for example. But here it's not so important as we can clone all args. And in the current state of the code, we don't make use of the returned resources anyway, so no harm in removing that. But generally speaking, I would argue this is good practice and a patterns that can be useful to know in Rust code.
- Personally, I only create error enums when there is more than one failure case. I guess it's mostly a matter of taste. But my main argument would be that refactoring from an implicit error to a enum error is trivial when a second variant becomes necessary. But going back can be tricky if the type gets integrated into other types, for example as #[source] of another error. Also, going back to bullet point 1, returning resources is awkward when using a specific error type.

But again, looking at your code changes, it has made things more readable. So in this case I am in favor of making the change. :+1:

File: runtime/runtime/src/prefetch.rs (hunk 13 above)
Comment (pugachAG) [reply runtime/runtime/src/prefetch.rs:228]: ah, indeed, so I guess imports above also violate this, so I'm changing those as well

File: core/store/src/trie/prefetching_trie_storage.rs [hunk 15]
```diff
@@ -415,13 +432,12 @@ impl PrefetchApi {
...
 
-    /// Returns the argument back if queue is full.
     pub fn prefetch_trie_key(
         &self,
         root: StateRoot,
         trie_key: TrieKey,
-    ) -> Result<(), (StateRoot, TrieKey)> {
-        self.work_queue_tx.try_send((root, trie_key)).map_err(|e| e.into_inner())
+    ) -> Result<(), PrefetchError> {
+        self.work_queue_tx.try_send((root, trie_key)).map_err(|_| PrefetchError::QueueFull)
```
[the remark this answers is not in the document]
Comment (pugachAG) [reply core/store/src/trie/prefetching_trie_storage.rs]: Good point, I've added proper error handling here and a separate prefetch error for disconnect.

File: runtime/runtime/src/prefetch.rs [hunk 16]
```diff
@@ -157,23 +164,31 @@ impl TriePrefetcher {
...
+        match res {
+            Err(PrefetchError::QueueFull) => {
+                self.prefetch_queue_full.inc();
+                debug!(target: "prefetcher", "I/O scheduler input queue is full, dropping prefetch request");
+            }
+            Err(PrefetchError::QueueDisconnected) => {
+                // This shouldn't have happened, hence logging warning here
+                warn!(target: "prefetcher", "I/O scheduler input queue is disconnected, dropping prefetch request");
+            }
+            Ok(_) => self.prefetch_enqueued.inc(),
```
Comment (akhi3030) [inline runtime/runtime/src/prefetch.rs:178]: This makes sure that if the type changes, the compiler will tell us all the places we need to handle the situation.

```suggestion
            Ok(()) => self.prefetch_enqueued.inc(),
```

**Review by akhi3030 — APPROVED** [review]

File: runtime/runtime/src/prefetch.rs [hunk 17]
```diff
@@ -157,23 +164,27 @@ impl TriePrefetcher {
...
-        let queue_full = self.prefetch_api.prefetch_trie_key(self.trie_root, trie_key).is_err();
-        if queue_full {
-            self.prefetch_queue_full.inc();
-            debug!(target: "prefetcher", "I/O scheduler input queue full, dropping prefetch request");
-            Err(())
-        } else {
-            self.prefetch_enqueued.inc();
-            Ok(())
-        }
+    fn prefetch_trie_key(&self, trie_key: TrieKey) -> Result<(), PrefetchError> {
```
[the remark this answers is not in the document]
Comment (pugachAG) [reply runtime/runtime/src/prefetch.rs:167]: Just to confirm my understanding: you would like to avoid all `.clone()` calls involved in creating `TrieKey` from the main thread, is that right?

[the remark this answers is not in the document]
Comment (akhi3030) [reply runtime/runtime/src/prefetch.rs:167]: I feel like an API that encourages one to `clone()` a large data structure is problematic.  So we should think hard about it.  In this case, I wonder if changing the function to take `AccountId` will make the API less "error prone" to use.
